infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogFiles.java

```diff
@@ -275,13 +306,15 @@ private LogData readRecord(FileHandle fh, long address)
                     }
                 }
 
-                if (address == -1) {
+                if (isOverwritten(fh, filePosition)) {
+                    log.debug("Trim at {} hit, reading (size={}), position {}.", address, metadata.getLength(), filePosition);
+                } else if (address == -1) {
                     //Todo(Maithem) : maybe we can move this to getChannelForAddress
-                    fh.knownAddresses.add(entry.getGlobalAddress());
+                    fh.knownAddresses.put(entry.getGlobalAddress(), filePosition);
```

Comment:
Yes, please double check:

The only that could look like an issue IMHO is:

  readOverwrittenPositions(fh);
   readRecord(fh, -1);

Here you could ask - what if between these 2 new record is overwritten? readRecord() will return the old value. But as we are in a different thread and we don't break the sequential consistency, I don't think it is an issue.

Otherwise you see I hold the same write lock as for the segment.